x-pack/plugins/monitoring_collection/server/plugin.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { JsonObject } from '@kbn/utility-types';
+import { CoreSetup, CoreStart, Plugin, PluginInitializerContext, Logger } from 'kibana/server';
+import { registerDynamicRoute } from './routes';
+import { INDEX_TEMPLATE_KIBANA } from '../../monitoring/common/constants';
+import { MakeSchemaFrom } from '../../../../src/plugins/usage_collection/server';
+import { verifyMappings } from './lib';
+
+export interface MonitoringCollectionSetup {
+  registerMetric: <T>(metric: Metric<T>) => void;
+  getMetrics: <T>() => Promise<Record<string, MetricResult<T> | Array<MetricResult<T>>>>;
+}
+
+export type MetricResult<T> = T & JsonObject;
+
+export interface Metric<T> {
+  type: string;
+  schema: MakeSchemaFrom<T>;
+  fetch: () => Promise<MetricResult<T> | Array<MetricResult<T>>>;
+}
+
+export class MonitoringCollectionPlugin implements Plugin<MonitoringCollectionSetup, void, {}, {}> {
+  private readonly initializerContext: PluginInitializerContext;
+  private readonly logger: Logger;
+
+  private metrics: Array<Metric<any>> = [];
+  private disabledMetricTypes: string[] = [];
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.initializerContext = initializerContext;
+    this.logger = initializerContext.logger.get();
+  }
+
+  async getAllMetrics() {
```

Comment:
Is this only a function that SM internal collection would need to use?

Reply:
That's the current thinking, yes